### PR TITLE
RD-5753 Migrating app/utils files to TS - Part 1

### PR DESCRIPTION
**Pull request by Vorbert-Kruk**
## Description

Changes consists of migrating the following files:
- `app/utils/External.ts` 
- `app/utils/Manager.ts`

In the upcoming PR the rest of not migrated `app/utils/` files will be migrated 🎉 

## Screenshots / Videos

N/A

## Checklist

- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I followed the [UI Testing Guidelines](https://cloudifysource.atlassian.net/l/cp/udCVfvrx) and verified that all tests/checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests

 

N/A

## Documentation

N/A

### CODE REVIEW

File: app/utils/Manager.ts [hunk 1]
```diff
@@ -33,52 +43,59 @@ export default class Manager extends Internal {
...
             return fullData;
         });
     }
 
-    doPostFull(url, params, body) {
+    doPostFull<ResponseBody, RequestQueryParams extends QueryStringParams>(
+        url: string,
+        body: ResponseBody,
+        params?: RequestQueryParams
+    ) {
```
Comment (qooban) [inline app/utils/Manager.ts:94]: I see arguments order change, so no need to update any other places? 🤔

File: app/utils/Manager.ts [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
+    metadata: {
+        pagination: {
+            offset: 0,
+            size: 0,
+            total: 0
+        }
+    }
+};
 
 export default class Manager extends Internal {
```
Comment (qooban) [inline app/utils/Manager.ts:21]: Documentation:

I see inconsistency between `app/utils/Manager.ts` and https://docs.cloudify.co/staging/dev/developer/writing_widgets/widget-apis/#getmanager, but as I'll be doing more changes here, I'll take it as part of RD-6121.

File: app/utils/Manager.ts (hunk 1 above)
Comment (Vorbert-Kruk) [reply app/utils/Manager.ts:94]: Yeah, you're right 😅 
I've found one place needed an update - which has ben provided within 0da49adee992d10314c25310baf7325e6d8ceda3 🎉

File: app/utils/Manager.ts (hunk 2 above)
Comment (Vorbert-Kruk) [reply app/utils/Manager.ts:21]: Ou, that's good to know!
Till now I didn't know that we are documenting some of our inner classes, like e.g. `Manager`, `External` - new day, new knowledge portion 🥳 

Just out of curiosity, you're referring to the list of `Available methods for getting Cloudify Manager properties`?
Or do you see some inconsistency in other description, related to the `Manager`? 🤔

File: app/utils/Manager.ts [hunk 3]
```diff
@@ -33,52 +43,58 @@ export default class Manager extends Internal {
...
                 return this.doFetchFull(fetcher, fetchParams, fullData, cumulativeSize);
             }
+
             return fullData;
         });
     }
 
-    doPostFull(url, params, body) {
-        return this.doFetchFull(currentParams => this.doPost(url, { params: currentParams, body }), params);
+    doPostFull<ResponseBodyItem>(url: string, body: Record<string, any>, params?: QueryStringParams) {
```
Comment (qooban) [inline app/utils/Manager.ts:90]: Optional:

As:
```
    doPost<ResponseBody = any, RequestBody = any, RequestQueryParams extends QueryParams = QueryParams>(
        url: string,
        requestOptions?: RequestOptions<RequestBody, RequestQueryParams>
    ) {
        return this.ajaxCall<ResponseBody, RequestBody, RequestQueryParams>(url, 'post', requestOptions);
    }
```
I assume that `body` (the second argument of `doPostFull`) don't have to be an object (`Record<string, any>`).

File: app/utils/Manager.ts [hunk 4]
```diff
@@ -33,52 +43,58 @@ export default class Manager extends Internal {
...
+        const path = `/sp${getUrlWithQueryString(url, params)}`;
         return StageUtils.Url.url(path);
-    }
-
-    doFetchFull(fetcher, params = {}, fullData = { items: [] }, size = 0) {
+    };
+
+    doFetchFull<ResponseBodyItem>(
+        fetcher: (params: QueryStringParams) => Promise<PaginatedResponse<ResponseBodyItem>>,
+        params: QueryStringParams = {},
```
Comment (qooban) [inline app/utils/Manager.ts:66]: Optional:

You could probably extend that generic type to get second one and be something like `doGet` in `External` class:

```suggestion
    doFetchFull<ResponseBodyItem, RequestQueryParams extends QueryParams = QueryParams>(
        fetcher: (params: RequestQueryParams) => Promise<PaginatedResponse<ResponseBodyItem>>,
        params: RequestQueryParams = {},
```

Similar approach can be taken to `doPostFull`.

I didn't test it. 

It's definitely not a blocker.

File: app/utils/Manager.ts (hunk 2 above)
Comment (qooban) [reply app/utils/Manager.ts:21]: > Till now I didn't know that we are documenting some of our inner classes, like e.g. `Manager`, `External` - new day, new knowledge portion

Yeah, that might be something to improve. I mean - I'd like to make it clearer in the code - what parts are exposed and what parts are not.

> Just out of curiosity, you're referring to the list of Available methods for getting Cloudify Manager properties?

Yes, exactly to that part 👍🏻

**Review by qooban — APPROVED** [review]

File: app/utils/Manager.ts (hunk 2 above)
Comment (Vorbert-Kruk) [reply app/utils/Manager.ts:21]: Thanks for the explanation 🎉

File: app/utils/Manager.ts (hunk 4 above)
Comment (Vorbert-Kruk) [reply app/utils/Manager.ts:66]: As for now I resigned from the idea of passing a type for `params` (92f574de16f56908de28bce85331c503a1a0f3b5) I wouldn't push it within those changes, to keep things consistent 💪 

But idea seems to be interesting 🎉

File: app/utils/Manager.ts (hunk 3 above)
Comment (Vorbert-Kruk) [reply app/utils/Manager.ts:90]: Oh yeah, you're right - I'll adjust it in the upcoming PR 💪

File: widgets/common/src/actions/SearchActions.ts [hunk 5]
```diff
@@ -23,8 +25,11 @@ export default class SearchActions {
             .doPost(`/searches/${resourceName}`, { params, body: { filter_rules: filterRules } });
     }
 
-    private doListAll(resourceName: ResourceName, filterRules: FilterRule[], params?: Params) {
-        return this.toolbox.getManager().doPostFull(`/searches/${resourceName}`, params, { filter_rules: filterRules });
+    private doListAll<ResponseBody>(resourceName: ResourceName, filterRules: FilterRule[], params?: Params) {
+        return this.toolbox.getManager().doPostFull<ResponseBody>(`/searches/${resourceName}`, {
+            filter_rules: filterRules,
+            params
+        });
```
Comment (kubama) [inline widgets/common/src/actions/SearchActions.ts:32]: I suspect this change introduces a bug - before the change `params` were passed to `doPostFull` as the value for `params` argument, now they are part of the object that is passed as the value for `body` argument.

File: app/utils/Manager.ts [hunk 6]
```diff
@@ -33,52 +43,58 @@ export default class Manager extends Internal {
...
+        fullData: PaginatedResponse<ResponseBodyItem> = emptyPaginatedResponse,
```
Comment (kubama) [inline app/utils/Manager.ts:67]: Looks like this change introduces a bug.
Before the change the `fullData` argument was assigned a new object instance on each invocation without the argument value provided.
After the change the value assigned is always the same object for each invocation. This makes `emptyPaginatedResponse` to accumulate responses from different requests.

**Pull request merged**